packages/expo-screen-orientation/ios/ScreenOrientationRegistry.swift

```diff
@@ -0,0 +1,239 @@
+  // Called when iOS sends and OrientationDidChangeNotification
+  @objc func handleDeviceOrientationChange(notification: Notification) {
+    let newScreenOrientation = UIDevice.current.orientation.toInterfaceOrientation()
+
+    interfaceOrientationDidChange(newScreenOrientation)
+  }
+
+  // Called when the device is rotated.
+  func interfaceOrientationDidChange(_ newScreenOrientation: UIInterfaceOrientation) {
+    if currentScreenOrientation == newScreenOrientation || newScreenOrientation == .unknown {
+      return
+    }
+
+    // checks if screen orientation should be changed after user rotated the device
+    if currentOrientationMask.contains(newScreenOrientation) {
+      // when changing orientation without changing dimensions traitCollectionDidChange isn't triggered so the event has to be called manually
+      if (newScreenOrientation.isPortrait && currentScreenOrientation.isPortrait)
+        || (newScreenOrientation.isLandscape && currentScreenOrientation.isLandscape) {
+        screenOrientationDidChange(newScreenOrientation)
+        return
+      }
+
+      // on iPads, traitCollectionDidChange isn't triggered at all, so we have to call screenOrientationDidChange manually
+      if isIPad()
+        && (newScreenOrientation.isPortrait && currentScreenOrientation.isLandscape
+        || newScreenOrientation.isLandscape && currentScreenOrientation.isPortrait) {
+        screenOrientationDidChange(newScreenOrientation)
+      }
+    }
+  }
+
+  // Called by ScreenOrientationViewController when the dimensions of the view change
+  // Also used for Expo Go in EXAppViewController
+  @objc
+  public func traitCollectionDidChange(to traitCollection: UITraitCollection) {
+    currentTraitCollection = traitCollection
+
+    let verticalSizeClass = traitCollection.verticalSizeClass
+    let horizontalSizeClass = traitCollection.horizontalSizeClass
+    let currentDeviceOrientation = UIDevice.current.orientation.toInterfaceOrientation()
+    let currentOrientationMask = self.rootViewController?.supportedInterfaceOrientations ?? []
+
+    var newScreenOrientation = UIInterfaceOrientation.unknown
+
+    // We need to deduce what is the new screen orientaiton based on currentOrientationMask and new dimensions of the view
+    if verticalSizeClass == .regular && horizontalSizeClass == .compact {
+      // From trait collection, we know that screen is in portrait or upside down orientation.
+      let portraitMask = currentOrientationMask.intersection([.portrait, .portraitUpsideDown])
+
+      // Mask allows only proper portrait - we know that the device is in either proper portrait or upside down
+      // we deduce it is proper portrait.
+      if portraitMask == .portrait {
+        newScreenOrientation = .portrait
+      }
+      // Mask allows only upside down portrait - we know that the device is in either proper portrait or upside down
+      // we deduce it is upside down portrait.
+      else if portraitMask == .portraitUpsideDown {
```

Comment:
I think in that case it's better to put the comment inside the `if` block so we don't change/mix our formatting conventions.
```suggestion
      } else if portraitMask == .portraitUpsideDown {
        // Mask allows only upside down portrait - we know that the device is in either proper portrait or upside down
        // we deduce it is upside down portrait.
```